drivers/modbus_rtu/include/modbus_rtu.h

```diff
@@ -0,0 +1,173 @@
+#pragma once
+
+/**
+ * @defgroup    drivers_modbus ModBus
+ * @ingroup     drivers
+ * @brief       Modbus
+ *
+ * @{
+ * @file
+ * @brief       Modbus interface definitions
+ *
+ * @author      Denis Litvinov <li.denis.iv@gmail.com>
+ */
+
+#include "periph/uart.h"
+#include "thread.h"
+#include "mutex.h"
+
+/**
+ * @brief Modbus functions
+ */
+enum MB_FC {
+  MB_FC_NONE = 0,                /**< null operator */
+  MB_FC_READ_COILS = 1,          /**< FCT=1 -> read coils or digital outputs */
+  MB_FC_READ_DISCRETE_INPUT = 2, /**< FCT=2 -> read digital inputs */
+  MB_FC_READ_REGISTERS = 3,      /**< FCT=3 -> read registers or analog outputs */
+  MB_FC_READ_INPUT_REGISTER = 4, /**< FCT=4 -> read analog inputs */
+  MB_FC_WRITE_COIL = 5,          /**< FCT=5 -> write single coil or output */
+  MB_FC_WRITE_REGISTER = 6,      /**< FCT=6 -> write single register */
+  MB_FC_WRITE_COILS = 15,        /**< FCT=15 -> write multiple coils or outputs */
+  MB_FC_WRITE_REGISTERS = 16     /**< FCT=16 -> write multiple registers */
+};
+
+//! Максимальный размер пакета протокола Modbus RTU.
+#define MODBUS_RTU_PACKET_SIZE_MAX 256
+
+//! Размер контрольной суммы протокола Modbus RTU.
+#define MODBUS_RTU_CRC_SIZE 2
+
+//! Размер полей сообщения протокола Modbus RTU.
+#define MODBUS_RTU_FIELDS_SIZE (sizeof(modbus_rtu_id_t) + sizeof(modbus_rtu_func_t))
+//! Размер полей и контрольной суммы сообщения протокола Modbus RTU.
+#define MODBUS_RTU_FIELDS_CRC_SIZE (MODBUS_RTU_CRC_SIZE + MODBUS_RTU_FIELDS_SIZE)
+//! Максимальный размер данных протокола Modbus RTU.
+#define MODBUS_RTU_DATA_SIZE_MAX (MODBUS_RTU_PACKET_SIZE_MAX - MODBUS_RTU_FIELDS_CRC_SIZE)
+//! Максимальный размер данных и контрольной суммы протокола Modbus RTU.
+#define MODBUS_RTU_DATA_CRC_SIZE_MAX (MODBUS_RTU_DATA_SIZE_MAX + MODBUS_RTU_CRC_SIZE)
+
+/**
+ * @brief Modbus message structure.
+ */
+typedef struct {
+  uint8_t id;     /**< ID of slave, on response not use */
+  uint8_t func;   /**< function code */
+  uint16_t addr;  /**< Starting Address */
+  uint16_t count; /**< Quantity of Registers or byte or bits */
+  uint16_t *regs; /**< Registers or byte or bits*/
+} modbus_rtu_message_t;
+
+/**
+ * @brief Modbus structure.
+ */
+typedef struct {
+  /**
+   * @brief  UART port.
+   */
+  uart_t uart;
+  /**
+   * @brief  baudrate.
+   */
+  uint32_t baudrate;
+  /**
+   * @brief  RTS pin. 0 - disable.
+   */
+  gpio_t pin_rts;
+  /**
+   * @brief  RTS signals when transmit.
+   */
+  int pin_rts_enable;
```

Comment:
Would setting `pin_rts` to `GPIO_UNDEF` also work for disabling RTS. If so, this would save 2-4 bytes, depending on architecture.

Reply:
This mean
before transmit set `gpio_write(pin_rts, pin_rts_enable);`
after `gpio_write(pin_rts, !pin_rts_enable);`
Maybe this redundant, but in many library this parameter can be setup.

Reply:
You could add

```C
static inline int modbus_rtu_pin_enabled(modbus_rtu_t *modbus)
{
    return (modbus->pin_rts == GPIO_UNDEF) ? 0 : 1;
}
```

And than use `modbus_rtu_pin_enabled(modbus)` instead of `modbus->pin_rts_enable`. To me, this is equally readable, but the former safes 4 bytes of RAM.

Reply:
It used in 2 function. I`m think this not necessary.